### ISSUE TITLE
Support webhooks

### ISSUE THREAD

**Issue by mathstuf** [mgattozzi/github-rs#17]
Support for webhooks would be nice to have. Supporting deserializing all webhooks in a simple way can be done the same way I'm doing it for Gitlab [here](https://gitlab.kitware.com/utils/rust-gitlab/tree/f174c578933539c6daf482729cff65a14e46f492).

**Comment by mgattozzi** [comment]
Definitely! Webhooks is a pretty big thing considering that's often how the API is used. It's definitely a goal. PRs are welcome but chances are I'll have to implement this myself when I get the time. I can prioritize this over other portions of the API

**Comment by mathstuf** [comment]
No need to change priorities at the moment.

We'll probably need github support in the `ghostflow-director` at some point, but it hasn't formally hit the roadmap yet. This is the more active of the existing github crates, so I'm adding issues for what I know will eventually be necessary for us.

If you don't get around to the bits we'll need when we finally get around to it, I'll have PRs :) .

**Comment by mgattozzi** [comment]
That sounds good I'd be more than happy to review, merge, and then publish a new version so you can use it in your projects without a git dependency. Please send some in if I don't get around to it! Curious do you mean [this ghostflow-director](https://gitlab.kitware.com/utils/ghostflow-director) at all since it's also a Rust project?

**Comment by mathstuf** [comment]
Yep, that's the one. Well, one of them. The [rust-ghostflow crate](https://gitlab.kitware.com/utils/rust-ghostflow/) has the mechanisms, the director holds the policy decisions.

**Comment by mgattozzi** [comment]
Ahhh okay got it! Either way feel free to open up issues/PRs as needed. I'm more than happy to resolve these issues as they come up!

**Comment by mgattozzi** [comment]
I'll probably be working on this issue in particular more soon as I've started working on a Github Bot

**Comment by mathstuf** [comment]
Out of curiosity, what kind of bot? :)

**Comment by mgattozzi** [comment]
Something like bors or homu, hopefully making it more full featured to help automate tasks on GitHub. Was thinking of building it out with something like Tokio as a learning experience

**Comment by mathstuf** [comment]
Ah, we've built one for Gitlab (that's ghostflow-director), but alternative hosts are abstracted out (which is why I'm here :) ). The design is that it actually "listens" to jobs arriving through a directory dropped by a generic webhook-listen project (this is because hosts tend to have problems delivering hooks out-of-order due to exponential backoff when restarting the listener or it chokes on something). It also allows us to "replay" events easily by copying files back into the queue directory. Currently no async, but I have progress on getting `futures` to work, but there are some issues there that I need to work out (mainly we need some locking between jobs for ordering, but not all jobs conflict).

**Comment by mgattozzi** [comment]
Oh that's pretty neat actually!

**Comment by creekorful** [comment]
Hello there. I have face the same use case where I want to build a Github Bot in Rust.
Any news on this? Someone plan to integrate webhooks? Are help wanted?

Thanks in advance!
